src/ts/Query.ts

```diff
@@ -0,0 +1,305 @@
+import { InvalidParameterError } from '#shared/errors';
+import { BaseScalarArrayType, BaseScalarType } from '#shared/schemaType';
+import type { DbDocument, GenericObject } from '#shared/types';
+import type { ModelConstructor } from './compileModel';
+
+// #region Types
+export interface QueryConstructorOptions {
+	/** Skip the first _n_ results */
+	skip?: number;
+	/** Return only _n_ results */
+	limit?: number;
+	/** Sort criteria */
+	sort?: SortCriteria;
+	/** Return only the indicated properties */
+	projection?: string[];
+}
+
+export interface FilterOperators<TValue> {
+	/** Equal */
+	$eq?: TValue;
+	/** Greater than */
+	$gt?: TValue;
+	/** Greater than or equal to */
+	$gte?: TValue;
+	/** Less than */
+	$lt?: TValue;
+	/** Less than or equal to */
+	$lte?: TValue;
+	/** Not equal */
+	$ne?: TValue;
+	/** String containing */
+	$contains?: TValue;
```

Comment:
Does it make sense to type $contains, $startsWith, and $endWith as string?

Reply:
Of the scalar types, I can think of `string` and `number` as usable with these operators, although I would think that there would be a very limited use case for `number`.  The date types and the boolean type wouldn't make any sense.

I'm not sure its correct to change this here.  It's a little awkward in general that you _could_ use those operators on a property which isn't defined as a string at all.  It might be more awkward to force someone to provide a string for something that isn't a string were they to try to use the operator.

I did have a thought here though after looking at the MongoDB node types.  Perhaps these should be `TValue extends string ? string : never;`  That would probably be the best path because it would prevent a user from using it wrong.  I'll make that change.

![image](https://user-images.githubusercontent.com/4583869/157484732-5fbfbe12-937a-44ec-9746-10ebce8b0dd1.png)